tests/test_butterworth_filter.py

```diff
@@ -0,0 +1,192 @@
+
+    filter_arguments["cutoff_frequency"] = (20, 30)
+    bp_filter = BandPassFilter(**filter_arguments)
+    bs_filter = BandStopFilter(**filter_arguments)
+
+    for filter_ in (bp_filter, bs_filter):
+        check_filter_attributes(filter_, filter_arguments)
+
+    assert lp_filter.filter_type == "lowpass"
+    assert hp_filter.filter_type == "highpass"
+    assert bp_filter.filter_type == "bandpass"
+    assert bs_filter.filter_type == "bandstop"
+
+
+def test_specified_butterworth_equivalence(filter_arguments):
+    del filter_arguments["filter_type"]
+
+    filter1 = ButterworthFilter(**filter_arguments, filter_type="lowpass")
+    filter2 = LowPassFilter(**filter_arguments)
+
+    # filter1.__eq__(filter2) differs from filter2.__eq__(filter1)
+    assert filter1 == filter2
+    assert filter2 == filter1
+
+    filter3 = ButterworthFilter(**filter_arguments, filter_type="highpass")
+    filter4 = HighPassFilter(**filter_arguments)
+    assert filter1 != filter4
+    assert filter2 != filter4
+    assert filter3 == filter4
+
+    filter_arguments["cutoff_frequency"] = (20, 30)
+    filter5 = ButterworthFilter(**filter_arguments, filter_type="bandpass")
+    filter6 = BandPassFilter(**filter_arguments)
+    assert filter1 != filter6
+    assert filter2 != filter6
+    assert filter5 == filter6
+
+    filter7 = ButterworthFilter(**filter_arguments, filter_type="bandstop")
+    filter8 = BandStopFilter(**filter_arguments)
+    assert filter1 != filter8
+    assert filter2 != filter8
+    assert filter7 == filter8
+
+
+def test_butterworth_functionality():
+    sample_frequency = 50
+    freq_low = 1
+    freq_medium = 4
+    freq_high = 10
+    amplitude_medium = 0.5
+    amplitude_high = 0.1
+    lowpass_cutoff = (freq_low + freq_medium) / 2
+    highpass_cutoff = (freq_medium + freq_high) / 2
+
+    order = 4
+
+    t = np.arange(0, 100, 1 / sample_frequency)
+    low_part = np.sin(2 * np.pi * t * freq_low)
+    medium_part = np.sin(2 * np.pi * t * freq_medium)
+    high_part = np.sin(2 * np.pi * t * freq_high)
+    signal_ = low_part + amplitude_medium * medium_part + amplitude_high * high_part
+
+    def compare_filters(cutoff, filter_type, class_):
+        filter1 = ButterworthFilter(filter_type, cutoff, order, sample_frequency)
+        filter2 = class_(cutoff, order, sample_frequency)
+        result1 = filter1.apply_filter(signal_)
+        result2 = filter2.apply_filter(signal_)
+        assert np.array_equal(result1, result2)
+
+        b, a = signal.butter(order, cutoff, filter_type, fs=sample_frequency)
+        sp_result = signal.filtfilt(b, a, signal_)
+        assert np.array_equal(result1, sp_result)
+
+    compare_filters(lowpass_cutoff, "lowpass", LowPassFilter)
+    compare_filters(highpass_cutoff, "highpass", HighPassFilter)
+    compare_filters((lowpass_cutoff, highpass_cutoff), "bandpass", BandPassFilter)
+    compare_filters((lowpass_cutoff, highpass_cutoff), "bandstop", BandStopFilter)
```

Comment:
Maybe also check that the result has the same shape as the input and is different than the input?

Is it worthwhile testing at least one of these on a >1 dimensional object?

Reply:
I've added these suggestions in 1c9b82b4be1137bdddda148b6d648ae060be5f0e.